### PR TITLE
Implement `set_state` on `TaskInstancePydantic`

### PR DESCRIPTION
**Pull request by dstandish**
Small part towards AIP-44.

### CODE REVIEW

File: airflow/models/taskinstance.py [hunk 1]
```diff
@@ -1839,6 +1839,36 @@ def key(self) -> TaskInstanceKey:
...
         return TaskInstanceKey(self.dag_id, self.task_id, self.run_id, self.try_number, self.map_index)
 
+    @staticmethod
+    @internal_api_call
+    @provide_session
+    def _set_state(ti, state, session=NEW_SESSION):
+        TI = TaskInstance
+        if not isinstance(ti, TaskInstance):
+            ti = (
+                session.query(TI)
```
Comment (vincbeck) [inline airflow/models/taskinstance.py]: I dont know if it is really worth creating a variable for that

```suggestion
        TI = TaskInstance
        if not isinstance(ti, TaskInstance):
            ti = (
                session.query(TaskInstance)
```

**Review by potiuk — APPROVED** [review]

File: airflow/models/taskinstance.py (hunk 1 above)
Comment (vincbeck) [reply airflow/models/taskinstance.py]: If `not isinstance(ti, TaskInstance)` is `True` then what is the type of `ti`? Should not we just pass `task_id`, `dag_id`, `run_id` and `map_index` as parameters? I think the goal is that all AIP-44 related APIs have primary keys as parameter which we then use to fetch the entity (task instance, dag_run, ...)

Comment (dstandish) [reply airflow/models/taskinstance.py]: TaskInstancePydantic in that case

Comment (dstandish) [reply airflow/models/taskinstance.py]: > Should not we just pass task_id, dag_id, run_id and map_index as parameters?

ok sure

Comment (dstandish) [reply airflow/models/taskinstance.py]: wait...
no so the reason for this is, if we already have a TI object there is no need to go fetch it from the DB.

wdyt?

Comment (vincbeck) [reply airflow/models/taskinstance.py]: And if AIP-44 is enabled, what is sent through the wire? `TaskInstancePydantic` I guess, which is serializable. Makes sense to me

File: airflow/models/taskinstance.py [hunk 2]
```diff
@@ -1839,6 +1839,35 @@ def key(self) -> TaskInstanceKey:
         """Returns a tuple that identifies the task instance uniquely."""
         return TaskInstanceKey(self.dag_id, self.task_id, self.run_id, self.try_number, self.map_index)
 
+    @staticmethod
+    @internal_api_call
+    def _set_state(ti: TaskInstance | TaskInstancePydantic, state, session: Session) -> bool:
+        TI = TaskInstance
+        if not isinstance(ti, TaskInstance):
+            ti = (
+                session.query(TI)
```
Comment (Taragolis) [inline airflow/models/taskinstance.py]: [Query API](https://docs.sqlalchemy.org/en/20/orm/queryguide/query.html) is deprecated in SQLAlchemy, could we use [session.execute](https://docs.sqlalchemy.org/en/14/orm/session_api.html#sqlalchemy.orm.Session.execute) here ?

File: airflow/models/taskinstance.py (hunk 1 above)
Comment (dstandish) [reply airflow/models/taskinstance.py]: ok

**Review by vincbeck — APPROVED** [review]

**Review by eladkal — APPROVED** [review]

**Pull request merged**